refactor:  migrating to typescript

hanwencheng opened the pull request:
closes #418.

Include update to React Navigation v4.

This PR I avoid to make changes to all the functionalities and UI features.

But may need to test all the signing issues.

Other changes includes: 
* reorder the file structure
* use alias for import all `components`, `screens`, `styles`, `constants`, `utils`, `stores`, `types`
* update requirement of node to `>=10` because of dev dependency `fsevents`

Tbaut commented on jest.config.js:
```diff
@@ -14,14 +14,30 @@
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
-const { defaults } = require('jest-config');
+// const { defaults } = require('jest-config');
```
stray comment

hanwencheng replied:
good catch! also deleted other `console.log`

sjeohp commented:
Testing on iOS.

- Adding legacy account doesn't look right.

<img width="375" alt="Screenshot 2020-03-02 at 11 46 49 AM" src="https://user-images.githubusercontent.com/3927362/75670496-233ba400-5c7d-11ea-882e-e5abacc6a5ab.png">

- `<` button on `About` screen is sometimes invisible. Don't know how to reproduce reliably, it just happens sometimes.

![IMG_9E6793CB433C-1](https://user-images.githubusercontent.com/3927362/75670590-4fefbb80-5c7d-11ea-976d-1625e0933315.jpeg)

- For an Ethereum account: 
1. Choose `Delete`
2. Alert (`Do you really want to delete key pairs for this account`) sometimes appears, sometimes does not.
3. `Unlock Identity` pin entry fails (`You pin must be at least 6 digits long!`) even when the pin is correct.
4. `<` back button takes me to a blank screen.
![IMG_FA72A5713C1B-1](https://user-images.githubusercontent.com/3927362/75672153-6d725480-5c80-11ea-9d5d-fc48b6a0cdd3.jpeg)
5. `<` from the blank screen takes me to the accounts list and the Ethereum account has been deleted.

sjeohp commented:
- Account icons sometimes disappear, possibly after the buggy Ethereum deletion.
![IMG_6C25F77A1BC2-1](https://user-images.githubusercontent.com/3927362/75672358-e5d91580-5c80-11ea-8f1d-f2af11d32d7e.jpeg)

hanwencheng commented:
Re @sjeohp , thanks for checking it!

1. Add legacy account screen is only used for developing, it will never shows in production app, so I just use it to check backward-compatibility, I will leave it for now.

2. Ethereum deletion problem is a vital bug, I have fixed it, also noticed that the alert text with "key pairs" is not clear to the user, so delete all the occurence of "key pairs" and use "account" instead.

3. With checking back button issue, I tried hard but still can't reproduce it on either Andorid and iOS, probably it is related to ehtereum account deletion problem. I guess it is caused by react navigation, so I updated  it to latest version. 

4. When trying reproduce `3`, found a new bug, "after screen transition, only after the small delay, buttons on new screen will work."  it is fixed by upgrading react navigation. 

5. Network Icon missing problem, it is a bug by misusage of react `useEffect` hook, with async function in `useEffect`. The bug is updating UI (in `useEffect` hook) after the screen is destroyed. So I introduces `eslinit-plugin-react-hooks` into Eslint, and fixes the bug in both `AccoutIcon` and `QrView`.
